core/src/main/java/org/elasticsearch/index/query/IndicesQueryBuilder.java

```diff
@@ -31,48 +39,67 @@
 
     public static final String NAME = "indices";
 
-    private final QueryBuilder queryBuilder;
+    /** The default query to be executed on non matching indices */
+    public static final QueryBuilder DEFAULT_NO_MATCH_QUERY = QueryBuilders.matchAllQuery();
+
+    private final QueryBuilder innerQuery;
 
     private final String[] indices;
 
-    private String sNoMatchQuery;
-    private QueryBuilder noMatchQuery;
+    private QueryBuilder noMatchQuery = DEFAULT_NO_MATCH_QUERY;
 
     static final IndicesQueryBuilder PROTOTYPE = new IndicesQueryBuilder(null);
 
-    public IndicesQueryBuilder(QueryBuilder queryBuilder, String... indices) {
-        this.queryBuilder = queryBuilder;
+    public IndicesQueryBuilder(QueryBuilder innerQuery, String... indices) {
+        this.innerQuery = innerQuery;
         this.indices = indices;
     }
 
+    public QueryBuilder innerQuery() {
+        return this.innerQuery;
+    }
+
+    public String[] indices() {
+        return this.indices;
+    }
+
     /**
      * Sets the no match query, can either be <tt>all</tt> or <tt>none</tt>.
      */
     public IndicesQueryBuilder noMatchQuery(String type) {
-        this.sNoMatchQuery = type;
+        this.noMatchQuery = parseNoMatchQuery(type);
         return this;
     }
+    
```

Comment:
I think I would make a breaking change here. Let's drop support for the string value in the builder and add it to the breaking changes. The parser still supports `none` and `all` but the builder only accepts a query. Then the method below needs to pretty much be moved to the parser.